web/shared/src/typing_status.ts

```diff
@@ -41,12 +45,20 @@ function same_recipient(a: Recipient | null, b: Recipient | null): boolean {
         return false;
     }
 
-    if (a.message_type === "direct" && b.message_type === "direct") {
-        // direct message recipients
-        return _.isEqual(a.ids, b.ids);
-    } else if (a.message_type === "stream" && b.message_type === "stream") {
-        // stream recipients
-        return same_stream_and_topic(a, b);
+    if (
+        a.notification_event_type === "typing_message_edit" &&
+        b.notification_event_type === "typing_message_edit"
+    ) {
+        // message edit
+        return _.isEqual(a.id, b.id);
+    } else if (a.notification_event_type === "typing" && b.notification_event_type === "typing") {
+        if (a.message_type === "direct" && b.message_type === "direct") {
+            // direct message recipients
+            return _.isEqual(a.ids, b.ids);
+        } else if (a.message_type === "stream" && b.message_type === "stream") {
+            // stream recipients
+            return same_stream_and_topic(a, b);
+        }
```

Comment:
I'm not sure whether or not this logic makes sense. The previous code was reliant on the assumption that at most one message can be in the compose box at a time.

This new `typing_message_edit` event type involves a separate input box from the compose box, and so possibly should be treated as a parallel/independent input.

Reply:
So in case of `message editing` this method basically checks out if the new notification received, is it of the same recipient as that of the current one. If yes, then continue showing the editing notification.

So in case when you are editing one message, and before the notification of this expires, move to another message and start editing that, then by the current method, the notification of the previous message editing would be stopped, and the editing notification of current message would be shown.

Do you want the previous message editing notification to be still displayed along with the current message's notification until it expires naturally?

Reply:
I talked with @alya and she confirmed that the product model that makes sense for this is to just have typing notifications tracked separately for different inputs.

(It's certainly possible to have multiple compose/edit inputs open and with current work in them at the same time).

I think if you refactor to support that model, it'll be a lot easier to read/debug/verify than doing this overloading.

Reply:
I'll update it.